x/stake/tick_test.go

```diff
@@ -120,14 +127,522 @@ func TestProcessProvisions(t *testing.T) {
+		require.Equal(t, startTotalSupply+expProvisions, pool.TotalSupply)
+
+		//hour 1600 was arbitrarily picked to bond the largest candidate
+		if hr <= 1600 {
+			require.Equal(t, unbondedTokens, pool.UnbondedPool)
+		} else {
+			require.Equal(t, initialUnbondedTokens-cand9bondedTokens, pool.UnbondedPool)
+		}
+
+		//steps to create unbonding of a candidate at hour 1600, to increase ratio from ~55% to ~73%
+		if hr == 1600 {
+			candidate, found := keeper.GetCandidate(ctx, addrs[9])
+			assert.True(t, found)
+			beforeBondedRatio := pool.bondedRatio()
+
+			// bond 100,000,000 tokens that were previously unbonded
+			pool, candidate = pool.unbondedToBondedPool(candidate)
+			keeper.setPool(ctx, pool)
+			unbondedShares = unbondedShares.Sub(unbondSharesCand9)
+
+			//candidate.Assets are the shares. shares should be less than 100,000, bondedShareExRate() should be greater than 1, when multiplied they equal 100,000
+			cand9bondedTokens = pool.bondedShareExRate().Mul(candidate.Assets).Evaluate()
+			cand9unbondedTokens = cand9unbondedTokens - cand9bondedTokens
+			assert.Equal(t, int64(100000000), cand9bondedTokens)
+
+			// must add cumulativeExpProvs here, to get true bonded tokens at this instance, to find new value for bondedShares
+			bondedAt1600 := bondedTokens + cand9bondedTokens + cumulativeExpProvs
+
+			// bonded shares should increase
+			bondedShares = sdk.NewRat(bondedAt1600, 1).Quo(pool.bondedShareExRate())
+			assert.True(t, bondedShares.GT(sdk.NewRat(300000000, 1)))
+
+			//Ensure that new bonded ratio is greater than old bonded ratio, since we just added 100,000 bonded
+			assert.True(t, (pool.bondedRatio().GT(beforeBondedRatio)))
+		}
+	}
+
+	// Final checks after 8766 hours to ensure pool has correct values
+	pool = keeper.GetPool(ctx)
+	calculatedTotalTokens := totalSupply + cumulativeExpProvs
+	calculatedBondedTokens := bondedTokens + cumulativeExpProvs + cand9bondedTokens
+	calculatedUnbondedTokens := calculatedTotalTokens - calculatedBondedTokens
+	assert.NotEqual(t, initialTotalTokens, pool.TotalSupply)
+
+	//make sure bonded ratio is correct
+	assert.True(t, pool.bondedRatio().Equal(sdk.NewRat(calculatedBondedTokens, calculatedTotalTokens)), "%v", pool.bondedRatio())
+
+	// global token supply checks
+	assert.Equal(t, calculatedTotalTokens, pool.TotalSupply)
+	assert.Equal(t, calculatedBondedTokens, pool.BondedPool)
+	assert.Equal(t, unbondedTokens-cand9bondedTokens, pool.UnbondedPool)
+
+	// test the value of candidate shares. bonded ex rate should be greater than 1, unbonded should still be 1
+	assert.True(t, pool.bondedShareExRate().Mul(bondedShares).Equal(sdk.NewRat(calculatedBondedTokens)), "%v", pool.bondedShareExRate())
+	assert.True(t, pool.unbondedShareExRate().Mul(unbondedShares).Equal(sdk.NewRat(calculatedUnbondedTokens)), "%v", pool.unbondedShareExRate())
+}
+
+//Tests that inflation works as expected when we get a randomly updating sample of candidates
+func TestAddingRandomCandidates(t *testing.T) {
```

Comment:
Overall this test is quite difficult to read/understand - we should break it out into many small functions which are called by an overall concise Test.